### DISCUSSION TITLE
PyInstaller and UPX on Linux -  No change in file size with or without UPX

### DISCUSSION THREAD

**Discussion by Aditya-ds-1806** [pyinstaller/pyinstaller#6275] [Help] [answered]
I am trying to use PyInstaller and UPX to bundle a flask application on Ubuntu. When I run:
```bash
pyinstaller app.py --onefile --clean
```
the logs show that UPX is available:
```bash
297 INFO: UPX is available.
```
However, in the entire log file, there doesn't seem to be a single instance of PyInstaller calling UPX. Consequently, the file size remains same irrespective of whether UPX is installed or not. I did not face this issue on Windows with the same command as above. The docs say:

> PyInstaller looks for UPX on the execution path or the path specified with the `--upx-dir` option. If UPX exists, PyInstaller applies it to the final executable, unless the `--noupx` option was given.

Clearly I don't need to specify `--upx-dir` since PyInstaller was able to find UPX. But why isnt it getting applied then? What is going wrong?

**Comment by rokm** [comment, marked as answer]
You're right, UPX is actually used only on Windows. See #5966.

**Reply by Aditya-ds-1806** [reply]
So there is no way to tell pyinstaller to apply upx on linux? @rokm

**Reply by bwoodsend** [reply]
No but you can apply it manually and discover why there is no option for it in PyInstaller.

**Reply by rokm** [reply]
> So there is no way to tell pyinstaller to apply upx on linux? @rokm

No, it is permanently disabled here:
https://github.com/pyinstaller/pyinstaller/blob/5a02f55c696f16b98f23a8b487f3daa8f644a8d2/PyInstaller/building/utils.py#L182-L183

You can try enabling it, but unless you're building on old linux distribution, it will likely result in a defunct application (which will segfault as soon as it will try to dlopen the upx-processed python shared library).

**Reply by Avasam** [reply]
Has an issue been raised on UPX' side? I can't find anything on their issue tracker concerning Python shared libs on Linux.

**Reply by rokm** [reply]
> Has an issue been raised on UPX' side? I can't find anything on their issue tracker concerning Python shared libs on Linux.

Not by us - we have enough other issues to keep us busy...

FWIW, since https://github.com/pyinstaller/pyinstaller/pull/7707, you can force-enable UPX by setting `PYINSTALLER_FORCE_UPX` environment variable to 1 before running PyInstaller. You're expected to deal with the fallout yourself, though.

Contemporary versions of UPX on bleeding-edge linux distributions (I have 5.0.2 on my Fedora 42 system) seem to raise `CantPackException: DT_GNU_HASH above stub` on pretty much all shared libs collected from the system, so a simple "hello world" sees no difference is size. Suppose this is related to https://github.com/upx/upx/issues/625#issuecomment-1314720877 (and other issues linked in discussion there).

**Reply by bwoodsend** [reply]
UPX on Arch (also 5.0.2) seems to work well for me now. Building `print("hello")` with `--strip` then running `upx --lzma dist/test/_internal/*.so.*` goes from 14 MB to 7.5 MB. If I get PyInstaller to run upx, it does so after running strip which sounds like a bad idea to me and brings the size back up to 12 MB.

I guess if we can make UPX opt-in rather than opt-out (https://github.com/pyinstaller/pyinstaller/pull/8653) then we no longer need to gatekeep who can run UPX safely.

**Reply by rokm** [reply]
> UPX on Arch (also 5.0.2) seems to work well for me now.

Seems to also work for Debian Bookworm. I guess `CantPackException: DT_GNU_HASH above stub` might be a Fedora thing for now, but it may also be an indicator of things to come in other distributions.

> I guess if we can make UPX opt-in rather than opt-out (https://github.com/pyinstaller/pyinstaller/pull/8653) then we no longer need to gatekeep who can run UPX safely.

True. (Although the primary motivation there is to ensure it is opt-in on Windows). But now I'm starting to think that we should probably also make it explicit opt-in for existing spec files - for example by renaming the `upx` option into something like `use_upx` and emitting a warning if the old `upx` is set.

**Reply by Avasam** [reply]
> Not by us - we have enough other issues to keep us busy...

No worries, I was mainly wondering if I had missed something before trying out this super simple reproducer https://github.com/orgs/pyinstaller/discussions/8922#discussioncomment-11571639 and opening an issue myself.

Although I also just read https://github.com/orgs/pyinstaller/discussions/8922#discussioncomment-13185670 , and as bwoodsend just mentioned, a simple hello world reproducer no longer triggers a segfault with latest UPX.

> > UPX on Arch (also 5.0.2) seems to work well for me now.
>
> Seems to also work for Debian Bookworm.

My use-case is debian-based anyway (if someone wants to use my Python project on Arch or Fedora, they're welcome to figure out the package installation, build, and IO permissions themselves...)

> since https://github.com/pyinstaller/pyinstaller/pull/7707, you can force-enable UPX by setting `PYINSTALLER_FORCE_UPX` environment variable to 1 before running PyInstaller. You're expected to deal with the fallout yourself, though.

Interesting, I might play around with it later. Reading that other discussion topic about UPX 5.0.1, I'm not expecting much given I'm using PySide (Qt) and `--onefile`, but I can try the one-folder mode and either way raise issues I find upstream. Even if nothing improves anytime soon, at least having somewhere to track known existing issues.

Slightly off-topic, but still relevant when raising segmentation fault issues to UPX: https://github.com/upx/upx/issues/911#issuecomment-3362426407
> SIGSEGV debugging help: All current Continuous Integration builds (see the Actions page [upx/upx/actions](https://github.com/upx/upx/actions)) for amd64, arm64, arm32 ELF main programs on Linux support the compress-time command-line parameter `--catch-sigsegv` which intercepts any SIGSEGV during de-compression, then uses `gdb` to print relevant information (pc, register contents, disassembled instruciton stream) onto stderr. The static cost is about 800 bytes of code. If the `--catch-sigsegv` option is used and a SIGSEGV occurs without the debug info being printed onto stderr, then the UPX de-compressor was not to blame.

**Reply by Avasam** [reply]
I've had to exclude a few directories (I didn't bother finding exactly which libraries were causing segmentation faults), but got a claimed 53% compression on 22 files. Resulting in a total reduction of `11.5 MiB` (`347.3 MiB` --> `335.8 MiB`)

```ps1
$soFilesToCompress = Get-ChildItem -Path ./dist/AutoSplit -Recurse -File -Filter '*.so*' |
  Where-Object {
    -not (
      # _internal/*.so* causes Segmentation fault
      $_.Directory -like '*/AutoSplit/_internal' -or
      # _internal/PySide6/Qt/*/*.so* causes Segmentation fault
      # _internal/PySide6/Qt/plugins/*/*.so* breaks style
      $_.Directory -like '*/AutoSplit/_internal/PySide6/Qt/*'
    )
  }
./scripts/.upx/upx --lzma --best ./dist/AutoSplit/AutoSplit $soFilesToCompress
```

Results:
```
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2026
UPX 5.1.1       Markus Oberhumer, Laszlo Molnar & John Reiser    Mar 5th 2026

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
   5157344 ->   5120396   99.28%   linux/amd64   AutoSplit                     
    120920 ->     61540   50.89%   linux/amd64   _multiarray_tests.cpython-314-x86_64-linux-gnu.so
upx: ./dist/AutoSplit/_internal/numpy/_core/_multiarray_umath.cpython-314-x86_64-linux-gnu.so: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/numpy/linalg/_umath_linalg.cpython-314-x86_64-linux-gnu.so: CantPackException: DT_GNU_HASH above stub
    510632 ->    189028   37.02%   linux/amd64   _pocketfft_umath.cpython-314-x86_64-linux-gnu.so
    306344 ->    103604   33.82%   linux/amd64   _bounded_integers.cpython-314-x86_64-linux-gnu.so
    238632 ->     93236   39.07%   linux/amd64   _common.cpython-314-x86_64-linux-gnu.so
    917304 ->    292676   31.91%   linux/amd64   _generator.cpython-314-x86_64-linux-gnu.so
    115992 ->     64804   55.87%   linux/amd64   _mt19937.cpython-314-x86_64-linux-gnu.so
    129080 ->     57412   44.48%   linux/amd64   _pcg64.cpython-314-x86_64-linux-gnu.so
    103704 ->     56612   54.59%   linux/amd64   _philox.cpython-314-x86_64-linux-gnu.so
     74808 ->     44100   58.95%   linux/amd64   _sfc64.cpython-314-x86_64-linux-gnu.so
    205496 ->     96964   47.19%   linux/amd64   bit_generator.cpython-314-x86_64-linux-gnu.so
    721656 ->    211716   29.34%   linux/amd64   mtrand.cpython-314-x86_64-linux-gnu.so
upx: ./dist/AutoSplit/_internal/PIL/_avif.cpython-314-x86_64-linux-gnu.so: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/PIL/_imaging.cpython-314-x86_64-linux-gnu.so: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/PIL/_imagingcms.cpython-314-x86_64-linux-gnu.so: CantPackException: DT_GNU_HASH above stub
     39320 ->     20900   53.15%   linux/amd64   _imagingmath.cpython-314-x86_64-linux-gnu.so
upx: ./dist/AutoSplit/_internal/PIL/_imagingtk.cpython-314-x86_64-linux-gnu.so: NotCompressibleException
upx: ./dist/AutoSplit/_internal/PIL/_webp.cpython-314-x86_64-linux-gnu.so: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/cv2/cv2.abi3.so: CantPackException: DT_GNU_HASH above stub
    433552 ->    218372   50.37%   linux/amd64   libshiboken6.abi3.so.6.9      
upx: ./dist/AutoSplit/_internal/shiboken6/Shiboken.abi3.so: NotCompressibleException
upx: ./dist/AutoSplit/_internal/pillow.libs/libavif-a883386a.so.16.4.1: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libjpeg-8296d2fa.so.62.4.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/liblcms2-545f6561.so.2.0.18: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/liblzma-3c942967.so.5.8.3: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libopenjp2-f3576751.so.2.5.4: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libsharpyuv-0bacc318.so.0.1.2: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libtiff-f4bc73ff.so.6.2.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libwebp-67ee3789.so.7.2.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libwebpdemux-eeb71312.so.2.0.17: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libwebpmux-20a90b09.so.3.1.2: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libXau-154567c4.so.6.0.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libxcb-bac6ebc8.so.1.1.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/pillow.libs/libzstd-6ea785c0.so.1.5.7: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libavcodec-76c43bf0.so.59.37.100: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libavformat-d296e685.so.59.27.100: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libavutil-734d06dd.so.57.28.100: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libcrypto-8c1ab3ad.so.1.1: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libgfortran-91cc3cb1.so.3.0.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libopenblas-r0-f650aae0.3.3.so: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libpng16-ef62451c.so.16.44.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libquadmath-96973f99.so.0.0.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libssl-28bef1ac.so.1.1: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libswresample-3e7db482.so.4.7.100: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libswscale-95ddd674.so.6.7.100: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/opencv_contrib_python_headless.libs/libvpx-9f572e11.so.9.1.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/numpy.libs/libgfortran-040039e1-0352e75f.so.5.0.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/numpy.libs/libquadmath-96973f99-934c22de.so.0.0.0: CantPackException: DT_GNU_HASH above stub
upx: ./dist/AutoSplit/_internal/numpy.libs/libscipy_openblas64_-8fb3d286.so: CantPackException: DT_GNU_HASH above stub
    455504 ->    129708   28.48%   linux/amd64   levenshtein_cpp.cpython-314-x86_64-linux-gnu.so
    242472 ->    146228   60.31%   linux/amd64   libpyside6.abi3.so.6.9        
   3846824 ->   1563316   40.64%   linux/amd64   QtCore.abi3.so                
    336136 ->    174356   51.87%   linux/amd64   QtDBus.abi3.so                
   4738184 ->   2126996   44.89%   linux/amd64   QtGui.abi3.so                 
   1225080 ->    530820   43.33%   linux/amd64   QtNetwork.abi3.so             
    206904 ->    106564   51.50%   linux/amd64   QtTest.abi3.so                
   5948984 ->   2534980   42.61%   linux/amd64   QtWidgets.abi3.so             
   --------------------   ------   -----------   -----------
  26074872 ->  13944328   53.48%                 [ 22 files ]

Packed 24 files: 22 ok, 2 errors.
```

Oh and for fun, here's the results with and w/o upx+strip:
<img width="542" height="537" alt="image" src="https://github.com/user-attachments/assets/e2b69bce-61d6-4d34-943d-78e89b62cb0b" />
